### ISSUE TITLE
🚀 Remote URL detection should parse `insteadOf` replacements

### ISSUE THREAD

**Issue by mathstuf** [dandavison/delta#693]
Git supports replacing parts of URLs with other strings. For example, I can give a remote URL of `github:user/project` and it will go off to `github.com` and clone it. If I try to push, it will instead use the `ssh` URL (this avoids having to set up different URLs for push and pull manually). I see that `delta` is parsing out the remote based on the `remote.$name.url` setting. However, it should really use `git remote get-url` to have `insteadOf` replaced as intended.

**Comment by dandavison** [comment]
Thanks @mathstuf! Is there any chance you could test whether https://github.com/dandavison/delta/pull/694 implements your suggestion? I haven't personally used this git feature, so am slightly unsure what standard usage patterns are. Similarly, if you could give me some example usages of a gitconfig entry making use of `insteadOf` then I'll use it to add a unit test for this feature.

[Compiling delta from source follows the standard pattern for any rust project: https://github.com/dandavison/delta#build-delta-from-source]

**Comment by dandavison** [comment]
The PR uses libgit2's [git_remote_url](https://libgit2.org/libgit2/#HEAD/group/remote/git_remote_url) which is documented to honor `insteadOf` substitutions.

**Comment by mathstuf** [comment]
I have this in my gitconfig:

```ini
[url "https://github.com/"]
    insteadOf = github:
[url "ssh://git@github.com/"]
    pushInsteadOf = github:
    insteadOf = githubpriv:
```

which allows me to do `git clone github:dandavison/delta` and `git clone githubpriv:mathstuf/myprivatefoo`.

**Comment by dandavison** [comment]
Thanks @mathstuf. Just to make sure I understand, can you state explicitly what the incorrect behavior in delta is? I'm thinking that you're reporting a bug regarding hyperlinks, i.e. that you have observed URLs being formatted incorrectly when using delta's `hyperlinks` feature -- is that correct? If so, can you tell me what your `[remote "origin"]` section in `./.git/config` looks like, and how the hyperlinks are being formatted currently, and what you expect?


E.g. here's my `[remote "origin"]` section:

```gitconfig
[remote "origin"]
    url = git@github.com:dandavison/delta.git
    fetch = +refs/heads/*:refs/remotes/origin/*
    gh-resolved = base
```

**Comment by mathstuf** [comment]
To be fair, I was grepping the source code looking for some things and came across the `GitHubRepo` variant and ended up looking back at how that got determined. I then saw the typical "ask the config for `remote.origin.url`" pattern. I authored `git remote get-url` for just this use case: to have a mechanism for tools to use when they encounter my `insteadOf` url settings.

Here is my settings for my clone of this repo:

```gitconfig
[remote "origin"]                                                                                                                                                                                                                          
        url = github:dandavison/delta                                                                                                                                                                                                      
        fetch = +refs/heads/*:refs/remotes/origin/*
```

Actually testing things, if I change it to `https://`, I can see the following using `tmux pipe-pane`:

```
^[[1;33mcommit ^[]8;;https://github.com/dandavison/delta/commit/ea3417abe4d7ab6399254263ec42aae2fb986512^[\ea3417abe4d7ab6399254263ec42aae2fb986512^[]8;;^[\ (HEAD -> master
```

otherwise all of the osc8 stuff is missing.

**Comment by dandavison** [comment]
Great, thanks for the extra explanation @mathstuf. I believe it's fixed in master by #694.
